File: 2020/day11.js
const fs = require('fs');

fs.readFile('input/day11', 'utf8', (_, data) => {
  const layout = data.trim().split('\n').map(l => l.split(''));

  console.log('Part 1:', count(layout));
  console.log('Part 2:', count(layout, true));
});

const count = (prev, rule2 = false, prevJSON = JSON.stringify(prev)) => {
  const state = nextRound(prev, rule2);
  const json = JSON.stringify(state);

  return json == prevJSON
    ? state.reduce((acc, row) => acc + row.filter(c => c == '#').length, 0)
    : count(state, rule2, json);
};

const nextRound = (layout, rule2) => layout.map((row, y) => {
  return row.map((cell, x) => {
    if (cell == '.') return '.';

    let neighboursCount = 0;

    for (let offX = -1; offX < 2; offX++) {
      for (let offY = -1; offY < 2; offY++) {
        if (offX == 0 && offY == 0) continue;

        neighboursCount += findSeat(layout, x, y, offX, offY, rule2);
      }
    }

    if (cell == 'L' && neighboursCount == 0) return '#';

    if (cell == '#' && neighboursCount >= 4 + rule2) return 'L';

    return cell;
  });
});

const findSeat = (layout, x, y, offsetX, offsetY, rule2) => {
  switch (layout[y + offsetY] && layout[y + offsetY][x + offsetX]) {
    case '.':
      return rule2
        ? findSeat(layout, x + offsetX, y + offsetY, offsetX, offsetY, rule2)
        : 0;

    case '#':
      return 1;
  }

  return 0;
};

File: 2020/day11.py
import itertools

def count(prev, rule2 = False):
    state = next_round(prev, rule2)

    if state != prev:
        return count(state, rule2)

    result = 0

    for row in state:
        for cell in row:
            if cell == "#":
                result += 1

    return result

def next_round(layout, rule2):
    result = list(map(lambda x: x[:], layout))

    for y, x in itertools.product(range(len(layout)), range(len(layout[0]))):
        neighbours_count = 0

        for off_x, off_y in itertools.product(range(-1, 2), range(-1, 2)):
            if off_x == 0 and off_y == 0:
                continue

            neighbours_count += find_seat(layout, x, y, off_x, off_y, rule2)

        if layout[y][x] == "L" and neighbours_count == 0:
            result[y][x] = "#"

        if layout[y][x] == "#" and neighbours_count >= 4 + int(rule2):
            result[y][x] = "L"

    return result

def find_seat(layout, x, y, off_x, off_y, rule2):
    cell = at(layout, x + off_x, y + off_y)

    if cell == "#":
        return 1

    if cell == "." and rule2:
        return find_seat(layout, x + off_x, y + off_y, off_x, off_y, rule2)

    return 0

def at(layout, x, y):
    if y in range(len(layout)) and x in range(len(layout[y])):
        return layout[y][x]

    return None

if __name__ == "__main__":
    file_contents = open("input/day11", "r").read()
    layout = list(map(list, file_contents.split()))

    print("Part 1:", count(layout))
    print("Part 2:", count(layout, True))
